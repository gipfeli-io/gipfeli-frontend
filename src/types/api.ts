export interface ApiResponseWrapper {
  success: boolean
  statusCode: number
  statusMessage: string
}

export interface SingleApiResponse<T> extends ApiResponseWrapper {
  content?: T
}

export interface ArrayApiResponse<T> extends ApiResponseWrapper {
  content?: T[]
}

export type ClassCastHint<T> = (new (...args: any[]) => T) | void
