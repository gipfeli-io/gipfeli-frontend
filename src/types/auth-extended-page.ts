import {NextComponentType, NextPage, NextPageContext} from 'next'
import {AppProps} from 'next/app'

export type NextPageWithAuth<P = {}, IP = P> = NextPage<P, IP> & {
    isPublic?: boolean
};

export type NextComponentWithAuth = NextComponentType<NextPageContext, any, {}> & Partial<NextPageWithAuth>

export type ExtendedAppProps<P = {}> = AppProps<P> & {
    Component: NextComponentWithAuth
};