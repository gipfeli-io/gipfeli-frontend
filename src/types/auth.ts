import { UserRole } from '../enums/user-role'

export class AuthObject {
  accessToken: string
  refreshToken: string

  constructor (accessToken: string, refreshToken: string) {
    this.accessToken = accessToken
    this.refreshToken = refreshToken
  }
}

/**
 * Basic JwtTokenPayload that is part of all JWTs.
 */
interface JwtTokenPayloadBase {
  iat: number;
  exp: number;
}

/**
 * In tokens, this UserIdentifier is the payload.
 */
export interface UserIdentifier {
  sub: string;
  email: string;
  role: UserRole;
}

/**
 * In tokens, this SessionIdentifier is the payload.
 */
export interface SessionIdentifier {
  sessionId: string;
}

/**
 * RefreshToken contains a session ID that maps to a UserSession object.
 */
export interface RefreshToken extends JwtTokenPayloadBase, SessionIdentifier {
}

/**
 * The main authtoken.
 */
export interface AccessToken extends JwtTokenPayloadBase, UserIdentifier {
}
