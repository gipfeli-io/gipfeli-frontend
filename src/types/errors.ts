import { ErrorType } from '../enums/error-type'

/**
 * This base error class is used to throw any non-specific errors that happen during any API calls. If not extending
 * NonCriticalApiError, these are handled by displaying a 500 error page.
 */
export class GenericApiError extends Error {
  constructor (message: string) {
    super(message)
    this.name = ErrorType.GENERIC
  }
}

/**
 * Statuscode: 500
 */
export class ServerError extends GenericApiError {
  constructor (message: string) {
    super(message)
    this.name = ErrorType.SERVER_ERROR
  }
}

/**
 * NonCriticalApiError are handled by triggering an error notification, but not interrupting the normal render flow.
 */
export abstract class NonCriticalApiError extends GenericApiError {
}

/**
 * Statuscode: 404
 */
export class NotFoundError extends NonCriticalApiError {
  constructor (message: string) {
    super(message)
    this.name = ErrorType.NOT_FOUND
  }
}

/**
 * Statuscode: 403
 */
export class ForbiddenError extends NonCriticalApiError {
  constructor (message: string) {
    super(message)
    this.name = ErrorType.FORBIDDEN
  }
}

/**
 * Statuscode: 401
 */
export class UnauthorizedError extends NonCriticalApiError {
  constructor (message: string) {
    super(message)
    this.name = ErrorType.UNAUTHORIZED
  }
}

/**
 * Displayed on login if either password or user or both are wrong
 */
export class LoginError extends NonCriticalApiError {
  constructor (message: string) {
    super(message)
    this.name = ErrorType.LOGIN_ERROR
  }
}
