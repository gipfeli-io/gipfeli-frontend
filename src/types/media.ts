import { GeoJsonObject, Point } from 'geojson'
import { GeometryObject } from './map'

/**
 * Base MediaObject - all media files have an ID and an identifier for the given storage.
 */
interface MediaObject {
  id: string;
  identifier: string;
}

/**
 * An uploaded image that might or might not have a location.
 */
export class ImageUpload extends GeometryObject implements MediaObject {
  id: string
  identifier: string
  location?: Point | null

  constructor (id: string, identifier: string, location: Point | null = null) {
    super()
    this.id = id
    this.identifier = identifier
    this.location = location
  }

  public getGeometry (): GeoJsonObject | null {
    return this.location ?? null
  }
}

export class GpxFileUpload implements MediaObject {
  id: string
  identifier: string
  name: string

  constructor (id: string, identifier: string, name: string) {
    this.id = id
    this.identifier = identifier
    this.name = name
  }
}

export interface UploadError {
  reason: string;
}
export interface CurrentUpload {
  name: string
  error?: UploadError
}
