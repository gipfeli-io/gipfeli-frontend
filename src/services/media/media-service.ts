import APIService from '../api-service'
import { RequestBody, SingleApiResponse } from '../../types/api'
import { GpxFileUpload, ImageUpload } from '../../types/media'

export default class MediaService extends APIService {
  private prefix: string = 'media'
  private fileField: string = 'file'

  constructor (token: string | undefined) {
    super()
    this.accessToken = token
  }

  public async uploadImage (image: File): Promise<SingleApiResponse<ImageUpload>> {
    return this.fetchSingleDataFromApi(
      this.getRequestUrl(this.prefix, 'upload-image'),
      this.getRequestBody('POST', image),
      ImageUpload
    )
  }

  public async uploadGpxFile (gpxFile: File): Promise<SingleApiResponse<GpxFileUpload>> {
    return this.fetchSingleDataFromApi(
      this.getRequestUrl(this.prefix, 'upload-gpx-file'),
      this.getRequestBody('POST', gpxFile),
      GpxFileUpload
    )
  }

  protected getRequestBody (method: 'POST', file: File): RequestBody {
    // Get the auth token header from the parent method, we'll override the content here for handling file uploads
    const fileUploadBody = super.getRequestBody(method)

    const content = new FormData()
    content.append(this.fileField, file)
    // Unset content-type to force fetch to set the content-type according to the uploaded file
    fileUploadBody.headers = { Authorization: fileUploadBody.headers.Authorization }
    fileUploadBody.body = content

    return fileUploadBody
  }
}
