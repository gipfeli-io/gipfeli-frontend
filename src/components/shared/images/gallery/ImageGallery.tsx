import React, { useEffect, useState } from 'react'
import { ImageUpload } from '../../../../types/media'
import { ImageList, ImageListItem } from '@mui/material'
import styles from './ImageGallery.module.scss'

type ImageGalleryProps = {
  images: ImageUpload[]
}

type ImageLink = {
  url: string;
  identifier: string;
}

const ImageGallery = ({ images }: ImageGalleryProps) => {
  const [imageLinks, setImageLinks] = useState<ImageLink[]>([])

  useEffect(() => {
    const mappedImages: ImageLink[] = images.map((item) => ({
      url: `${process.env.REACT_APP_STORAGE_BUCKET_BASE_URL + item.identifier}`,
      identifier: item.identifier
    }))

    setImageLinks(mappedImages)
  }, [images])

  const openInNewTab = (url: string): void => {
    const newWindow = window.open(url, '_blank', 'noopener,noreferrer')
    if (newWindow) newWindow.opener = null
  }

  const onClickUrl = (url: string): (() => void) => () => openInNewTab(url)

  return <>
    <ImageList cols={6} gap={8}>
      {imageLinks.map((item, index) => (
        <ImageListItem key={index}>
          <img
            className={styles.image}
            src={item.url}
            alt={item.identifier}
            loading="lazy"
            onClick={onClickUrl(item.url)}
          />
        </ImageListItem>
      ))}
    </ImageList>
  </>
}

export default ImageGallery
