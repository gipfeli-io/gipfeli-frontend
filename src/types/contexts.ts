import { Notification } from './notifications'
import { AppTheme } from './theme'
import { handleSave } from './handle-save'
import { CurrentUpload, GpxFileUpload, ImageUpload } from './media'
import { ConnectionStatus } from '../enums/connection-status'
import { ErrorInfo } from 'react'

export interface TourListContextProperties {
  deleteEvent: (id: string) => void
}

export interface ListContextProperties {
  deleteEvent: (id: string) => void
}

export interface AuthenticationContextType {
  email: string | undefined
  isAdmin: boolean
  token: string | undefined
  signIn: (email: string, password: string, callback: () => void) => Promise<void>
  signOut: (callback: () => void) => void
}

export interface NotificationContextType {
  notification: Notification | undefined
  triggerErrorNotification: (message: string) => void
  triggerSuccessNotification: (message: string) => void
  triggerOfflineNotification: () => void
  triggerSyncFailedNotification: (message: string) => void
  resetNotification: () => void
}

export interface ConnectionStatusContextType {
  isOffline: () => boolean
  updateConnectionStatus: (connectionStatus: ConnectionStatus) => void
  updateGoOnlineButtonVisibility: (isVisible: boolean) => void
  showGoOnlineButton: boolean
  isOnlineInfoBannerVisible: boolean
  updateOnlineInfoBannerVisibility: (isVisible: boolean) => void
  resetOnlineInfoBanner: () => void
}

export interface ThemeContextType {
  activeTheme: AppTheme
  toggleTheme: () => void
}

export interface ImageUploadContextType {
  files: ImageUpload[]
  save: handleSave<File[]>
  remove: (id: string) => void
  currentUploads: CurrentUpload[]
}

export interface GpxFileUploadContextType {
  file: GpxFileUpload
  save: handleSave<File>
  remove: (id: string) => void
  currentUpload?: CurrentUpload
}

export interface ErrorBoundaryContextType{
  triggerError: (error: Error, errorInfo?: ErrorInfo) => void
}
