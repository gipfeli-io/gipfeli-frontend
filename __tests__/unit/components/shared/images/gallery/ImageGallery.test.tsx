import '@testing-library/jest-dom'
import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import ImageGallery from '../../../../../../src/components/shared/images/gallery/ImageGallery'
import { ImageUpload } from '../../../../../../src/types/media'
import renderer from 'react-test-renderer'

const mockedStorageUrl = 'https://localhost:3000/'
const mockedImages: ImageUpload[] = [
  { id: 'img-1', identifier: 'ident-1' },
  { id: 'img-2', identifier: 'ident-2' }
]
describe('ImageGallery', () => {
  beforeEach(() => {
    process.env.REACT_APP_STORAGE_BUCKET_BASE_URL = mockedStorageUrl
  })
  it('renders a list of images with the correct url', () => {
    const { container } = render(
      <ImageGallery images={mockedImages}/>
    )

    const imgTags = container.querySelectorAll('img')

    expect(imgTags.length).toEqual(mockedImages.length)
    expect(imgTags[0].src).toEqual(`${mockedStorageUrl}${mockedImages[0].identifier}`)
    expect(imgTags[1].src).toEqual(`${mockedStorageUrl}${mockedImages[1].identifier}`)
  })

  it('behaves consistently', () => {
    const tree = renderer
      .create(<ImageGallery images={mockedImages}/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('has a click listener for opening a new window with the image url', () => {
    const windowSpy = jest.spyOn(window, 'open').mockImplementation(jest.fn())
    const { container } = render(
      <ImageGallery images={mockedImages}/>
    )
    const firstImage = container.querySelectorAll('img')[0]

    fireEvent.click(firstImage)

    const expectedParams = [
      `${mockedStorageUrl}${mockedImages[0].identifier}`,
      '_blank',
      'noopener,noreferrer'
    ]
    expect(windowSpy).toHaveBeenCalledTimes(1)
    expect(windowSpy).toHaveBeenCalledWith(...expectedParams)
  })
})
