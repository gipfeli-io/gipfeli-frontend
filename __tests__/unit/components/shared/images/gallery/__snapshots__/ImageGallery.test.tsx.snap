// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGallery behaves consistently 1`] = `
<ul
  className="MuiImageList-root MuiImageList-standard css-rvv586-MuiImageList-root"
  style={
    Object {
      "gap": 8,
      "gridTemplateColumns": "repeat(6, 1fr)",
    }
  }
/>
`;
